uqcsbot/dominos_coupons.py

```diff
@@ -139,27 +200,95 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
             for coupon in coupons
             if any(coupon.keyword_matches(keyword) for keyword in keywords)
         ]
+
+    random.shuffle(coupons)
+
     return coupons[:n]
 
 
-def _get_coupons_from_page() -> List[Coupon]:
+def _get_coupons_from_page(source: str) -> List[Coupon]:
```

Comment:
It would be nice to use the literal type again for `source`. This should avoid some of your case checking below. Additionally, I'd personally make this function only accept a single source (so not "both"), and let `_get_coupons` turn multiple sources into multiple calls of this function.